### ISSUE TITLE
N-Pair Loss reproducability

### ISSUE THREAD

**Issue by alpoler** [tjddus9597/Proxy-Anchor-CVPR2020#19]
Hi, I try to reproduce n-pair loss and i adhere to your hyperparameter but I can't get the result in the original paper. Which hyperparameter you use for n pair ? N pair loss in metric learning library is updated and some of your argument is not accepted. Also, could you share other parameter such as bone, emb learning rate, weight decay, batch size and image per class ?

**Comment by kdwonn** [comment]
Hello! N-pair was the hardest one to reproduce also in our case. We found out that using default PyTorch initialization(which means not using any initialization function) for the last FC layer is critical to reproduce reported performance. If you were using any initialization function (ex. nn.init.kaiming_normal), try without it. We used 1e-5 and 0.02 for the learning rate and l2-regularizer term weight.

**Comment by alpoler** [comment]
I'll change them. Thanks too much however original paper claim that all negative class appears in batch if it is possible but metric learning devise positive pair selection as choosing all unique combination in class balanced sampling scenario. Therefore, I consider that I should increase image per class to benefit diversity and number of pair to provide convergence. If it is possible, could you share sampling settings ?

**Comment by kdwonn** [comment]
For N-pair, we directly followed the batch construction setting described in the original paper. No sophisticated sampling technique was needed to reproduce it.  But I think tuning image-per-class could improve overall retrieval performance in some points. :smiley:

**Comment by alpoler** [comment]
I remove initialization and normalization also however l can't correct low accuracy although I apply hyperparameter of your trial. Do you add normalization in wrapping optimizer or explicitly from loss ? I use batch size of 144 with image per class 2 to get 72 pair like paper.

**Comment by kdwonn** [comment]
Can you give a detailed training setting Including learning rate scheduler, optimizer, weight decay, etc.?

**Comment by alpoler** [comment]
Lr scheduler stepLR ( STEP = 10 , DECAY_GAMA = 0.5)
Optimizer Adamw,weight_decay_0.0001,lr 1e-5 ( I add weight decay of 0.02 to embedder weights from optimizer or loss function)
batch size 144
IPC 2
l2-norm 0 
model resnet50
I cancel the weight initialize function in resnet50 class also like you mention.

**Comment by kdwonn** [comment]
~~Pair-based losses tend to converge slowly. Try with a larger LR decay step, like 100.~~
Also, note that our reproduction is based on BN-Inception. Hyperparameters and tricks for training might not work for other architectures.

OK, I just found the code I used a year ago.  You have to keep in mind the following things when you are working on N-pair loss.

- Do not freeze the batch-normalization layer when training on CUB, Cars.
- Use default initialization for the last FC layer.
- You don't have to use a learning rate scheduler.
- From my experience, Adam optimizer tends to works better than AdamW when using N-pair loss.

**Comment by alpoler** [comment]
Do you add regularization from loss explicitly or optimizer as embedding network weight decay  ? Thanks for parameter and your awesome work.

**Comment by kdwonn** [comment]
Do you mean 3.2.2 of the N-pair paper? L2 regularization term mentioned in the N-pair loss paper is a bit different from weight decay. Weight decay regularizes the norm of network weight to avoid overfitting, while the regularizer term regularizes the norm of embedding vector.  They can be used together since they do different things while training.

0.02 I mentioned is a weight multiplied to L2-regularization term in loss, which is for embedding vector. Check [pytorch_metric_learning.regularizers.LpRegularizer](https://kevinmusgrave.github.io/pytorch-metric-learning/regularizers/#lpregularizer). For weight decay, we used 5e-5.

**Comment by alpoler** [comment]
I constantly deal with proxies and mix the embedding vectors up with them. I realized that I should have performed it from loss to penalize embedding vectors after you say that. Are you sure about freezing batch normalization weights and bias ? Because it ruins all pretrained performance. Even I train 45 epochs for n pair, I get same around %2 percentage steadily . ( I forget to mention about embedding size also, it is 64)

**Comment by kdwonn** [comment]
Sorry for the late reply. Note that the training setting described in the README of this repo is only for Proxy-Anchor loss. While freezing BN is helpful for Proxy-Anchor and most of the losses, N-pair is not in the case.  I mentioned it already in a previous comment.

> ~Pair-based losses tend to converge slowly. Try with a larger LR decay step, like 100.~
> Also, note that our reproduction is based on BN-Inception. Hyperparameters and tricks for training might not work for other architectures.
> 
> OK, I just found the code I used a year ago. You have to keep in mind the following things when you are working on N-pair loss.
> 
> * Do not freeze the batch-normalization layer when training on CUB, Cars.
> * Use default initialization for the last FC layer.
> * You don't have to use a learning rate scheduler.
> * From my experience, Adam optimizer tends to works better than AdamW when using N-pair loss.

**Comment by kdwonn** [comment]
I'll close this issue for now. But please reopen it anytime if you have any additional questions. Thank you! :smile:
